apps/src/blockly/addons/cdoConstantsProvider.js

```diff
@@ -0,0 +1,122 @@
+import GoogleBlockly from 'blockly/core';
+import {inputBlockTypes, outputBlockStyleTypes} from './cdoConstants';
+
+export default class CdoConstantsProvider extends GoogleBlockly.blockRendering
+  .ConstantProvider {
+  /**
+   * Get an object with connection shape and sizing information based on the
+   * type of the connection.
+   *
+   * @param connection The connection to find a shape object for
+   * @returns The shape object for the connection.
+   * @override
+   */
+  shapeFor(connection) {
+    const inputTypeShapeMap = {
+      [inputBlockTypes.SPRITE]: this.TRI_INPUT_OUTPUT,
+      [inputBlockTypes.BEHAVIOR]: this.ROUND_INPUT_OUTPUT,
+      [inputBlockTypes.LOCATION]: this.RECT_INPUT_OUTPUT
+    };
+    const outputStyleShapeMap = {
+      [outputBlockStyleTypes.SPRITE_TYPE]: this.TRI_INPUT_OUTPUT,
+      [outputBlockStyleTypes.BEHAVIOR_TYPE]: this.ROUND_INPUT_OUTPUT,
+      [outputBlockStyleTypes.LOCATION_TYPE]: this.RECT_INPUT_OUTPUT
+    };
+    const blockStyleName = connection.getSourceBlock().styleName_;
+    const inputType = connection.check_ ? connection.check_[0] : null;
+    switch (connection.type) {
+      case GoogleBlockly.ConnectionType.INPUT_VALUE:
+        return inputTypeShapeMap[inputType] || this.PUZZLE_TAB;
+      case GoogleBlockly.ConnectionType.OUTPUT_VALUE:
+        return outputStyleShapeMap[blockStyleName] || this.PUZZLE_TAB;
+      case GoogleBlockly.ConnectionType.PREVIOUS_STATEMENT:
+      case GoogleBlockly.ConnectionType.NEXT_STATEMENT:
+        return this.NOTCH;
+      default:
+        throw Error('Unknown connection type');
+    }
+  }
+
+  makeTriangularInputConn() {
+    const width = this.TAB_WIDTH;
+    const height = this.TAB_HEIGHT;
+
+    /**
+     * Since input and output connections share the same shape you can
+     * define a function to generate the path for both.
+     */
+    function makeMainPath(up) {
+      return Blockly.utils.svgPaths.line([
+        Blockly.utils.svgPaths.point(-width, (-1 * up * height) / 2),
```

Comment:
btw I wonder if this should be `GoogleBlockly` here? I'm not sure exactly what `Blockly` refers to since you import `GoogleBlockly` in this file. This is just a random drive-by comment so feel free to ignore if it's not relevant.

Reply:
`Blockly` is a global variable we assign [here](https://github.com/code-dot-org/code-dot-org/blob/cb683e1000a3a9b491b60fb1f90cf1d27a4d3823/apps/src/sites/studio/pages/googleblockly.js#L7). And your comments are always welcome!

Reply:
Earlier, I tried replacing `GoogleBlockly` with `Blockly` but got the following errors:

![Screen Shot 2023-04-17 at 10 32 25 AM](https://user-images.githubusercontent.com/107423305/232536822-451b058f-cc41-4530-afdc-7775d2efe5dd.png)

![Screen Shot 2023-04-17 at 10 32 11 AM](https://user-images.githubusercontent.com/107423305/232536871-3cd2508e-1fb6-4ec6-93fe-07e71335d9f9.png)

I think that perhaps we need both instances - one that is customized by our wrapper and one that is mainline?

Reply:
@fisher-alice I think that both of these will have the same utilities so it maybe doesn't matter which we use... but if it's possible to use `GoogleBlockly.utils` without breaking anything, that feels a little tighter/cleaner. 
Currently, this just relies on that googleblockly.js script; it's not something we're likely to change but I guess it still counts as a dependency.

Reply:
Missed your earlier reply/screenshot. I can see that we might need to use a separate instance here (from global/window). If it breaks things, no need to change!

Reply:
Sorry late to the party here -- my immediate reaction is that I didn't think we should be referencing google blockly directly in our code -- we generally use the wrapper. We don't currently reference it directly anywhere else (unless we need it to subclass google blockly classes, as you are doing here), as far as I can tell. 

That said, thinking about it more, I do think that using it directly in these "addons" that are only referenced in the wrapper does feel less circular than trying to use the wrapper (which I'm pretty sure we do elsewhere, and maybe should not) -- since we only are using this code in the wrapper (via the different renderers). Does that make sense to folks? Happy to chat offline if what I'm saying isn't super clear.

Reply:
I don't have a strong opinion using `GoogleBlockly` vs the wrapper in the 'addons', but I do think we should be consistent, and you bring up a good point about how the wrapper is being used in other `addons'. 
@mikeharv , thoughts?

Reply:
After discussing offline, the group decided to reference `GoogleBlockly` considering comments above and also the eventual goal of using solely mainline and deprecating CDOBlockly.